iradix.go

```diff
@@ -357,6 +383,61 @@ func (t *Txn) delete(parent, n *Node, search []byte) (*Node, *leafNode) {
 	return nc, leaf
 }
 
+// delete does a recursive deletion
+func (t *Txn) deletePrefix(parent, n *Node, search []byte) (*Node, int) {
+	// Check for key exhaustion
+	if len(search) == 0 {
+		// Remove the leaf node
+		nc := t.writeNode(n, true)
+		subTreeSize := 0 //count this node
+		//recursively walk from all edges of the node to be deleted, tracking their mutate channels in the transaction
+		for _, e := range n.edges {
+			t.recursiveWatchNotify(e.node, func(n *Node) bool {
+				subTreeSize++
+				return false
+			})
+		}
+		if n.isLeaf() {
+			nc.leaf = nil
+			subTreeSize += 1
+		}
+		nc.edges = nil // deletes the entire subtree
+
+		return nc, subTreeSize
+	}
+
+	// Look for an edge
+	label := search[0]
+	idx, child := n.getEdge(label)
+	if child == nil || (!bytes.HasPrefix(child.prefix, search) && !bytes.HasPrefix(search, child.prefix)) {
```

Comment:
I think I'm finally understanding this a bit, though still going through some cases but i think the same prefix handling logic from `delete` should be sufficient, just the `len(search) == 0` case is different, as you've done.

```
         // Look for an edge
         label := search[0]
         idx, child := n.getEdge(label)
         if child == nil || !bytes.HasPrefix(search, child.prefix) {
                 return nil, nil
         }

         // Consume the search prefix
         search = search[len(child.prefix):]
         newChild, leaf := t.delete(n, child, search)
         if newChild == nil {
                 return nil, nil
         }
```

For the case you were worried about, if you delete "aca" prefix it will land on the node with the prefix "cat", consume "ca", and then call again with the empty search prefix, which will do the right thing. I don't think we could get down here for the other cases we are checking for in the PR, because we have to have the right prefix leading up to the node we are on. The only case that can happen is that the node has more stuff than our prefix, but that exhausts the search and terminates the algorithm.